Support Dynamic Peon Pod Template Selection in K8s extension

YongGang opened the pull request:
<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description
This PR introduces a new feature to dynamically select Kubernetes pod templates for task execution in Druid. This functionality aims to optimize resource utilization and improve task execution efficiency by tailoring pod specifications to the needs of different task characteristics.
Druid operator can define execution strategies and associate them with different task characteristics  through the new dynamic config interface. The system will apply these strategies dynamically as tasks are scheduled for execution.
This feature is a step towards making Apache Druid more adaptable and efficient in Kubernetes environments, addressing the need for more granular control over resource allocation and task scheduling.

#### Example Configuration:
We define two template keys in the configuration—`low-throughput` and `medium-throughput`—each associated with specific task conditions and arranged in a priority order.

- Low Throughput Template: This is the first template evaluated and has the highest priority. Tasks that have a context tag `billingCategory=streaming_ingestion` and a datasource of `wikipedia` will be classified under the `low-throughput` template. This classification directs such tasks to utilize a predefined pod template optimized for low throughput requirements.

- Medium Throughput Template: If a task does not meet the low-throughput criteria, the system will then evaluate it against the next selector in order. In this example, if the task type is index_kafka, it will fall into the `medium-throughput` template.

```
{
  "type": "default",
  "podTemplateSelectStrategy":
  {
    "type": "selectorBased",
    "selectors": [
      {
        "selectionKey": "low-throughput",
        "context.tags":
        {
          "billingCategory": ["streaming_ingestion"]
        },
        "dataSource": ["wikipedia"]
      },
      {
        "selectionKey": "medium-throughput",
        "type": ["index_kafka"]
      }
    ],
    "defaultKey": "base"
  }
}
```

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->

<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->
The Dynamic Pod Template Selection feature enhances the K8s extension by enabling more flexible and dynamic selection of pod templates based on task properties.

<hr>

##### Key changed/added classes in this PR
 * `KubernetesTaskRunnerDynamicConfig.java` Represents the configuration for task execution within a Kubernetes environment. This interface allows for dynamic configuration of task execution strategies based on specified behavior strategies.
 * `KubernetesTaskExecutionConfigResource.java` Resource that manages Kubernetes-specific execution configurations for running tasks.
 * `PodTemplateSelectStrategy.java` Defines a strategy for selecting Pod template of tasks based on specific conditions.
<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

arunramani commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/Selector.java:
```diff
@@ -0,0 +1,148 @@
+   */
+  @JsonCreator
+  public Selector(
+      @JsonProperty("name") String name,
```
Is `name` right here? It makes it feel like there can only be one of these for each category. Can we make it `categoryName` instead?

YongGang replied:
As this `Selector` class will be used for not only category but may be task laning work in the future, so I make it not tied to `categoryName` thus it can be reused easily for other purpose.

arunramani replied:
Ok, I think `name` is still a fairly confusing term because it seems like the name for the selector and not the category to select. Would a name like `selectionKey` or `selectKey` work better? That way it's general purpose enough.

YongGang replied:
updated with selectionKey

arunramani commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/PodTemplateTaskAdapter.java:
```diff
@@ -126,7 +131,17 @@ public PodTemplateTaskAdapter(
+      }
+    }
+    if (podTemplate == null) {
+      podTemplate = templates.getOrDefault(task.getType(), templates.get("base"));
```
It would be cool to make this another kind of behavior, and have it be default. That would be clean

georgew5656 commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/ExecutionConfig.java:
```diff
@@ -0,0 +1,43 @@
+})
+public interface ExecutionConfig
+{
+  String CONFIG_KEY = "k8s.taskrunner.config";
```
whats something that would go under this ExecutionConfig but not under ExecutionBehaviorStrategy? does it make more sense to call this KubernetesTaskRunnerRefreshableConfig or something like that?

YongGang replied:
`ExecutionConfig` can have config other than `ExecutionBehaviorStrategy`, we may can move `RunnerStrategy` to this dynamic config (or something similar in the future):
```
{
  "type": "default",
  "behaviorStrategy": {
    "type": "default",
    "categorySelectors": [
    ]
  },
  "runnerStrategy": {
    ...
  }
}
```
In this case it's not only guide `KubernetesTaskRunner` behavior but also whether task should run in Worker, so I think the general ExecutionConfig name is making more sense.

georgew5656 replied:
okay, ExecutionConfig makes sense to me then.

would categoryStrategy or templateStrategy make more sense for the second level thing then? since that config is really about choosing what template to run a task with?

georgew5656 replied:
would task laning not be a separate field? like laneStrategy?

YongGang replied:
I’m considering using a single strategy but with different fields returned. Since the implementation of the DynamicTaskExecutionBehaviorStrategy relies on the same Selector matching rules mechanism, the rules for matching categories and lanes are quite similar. Therefore, there’s no need to introduce a separate strategy for these functions.
```
public interface ExecutionBehaviorStrategy
{
  String getTaskCategory(Task task);
  String getTaskLane(Task task);
}
```

suneet-s replied:
YongGang and I discussed this offline, and I better understand the intent of this config object. It seems like this is trying to provide similar functionality as `KubernetesTaskRunnerConfig` but via the dynamic config. It makes sense to have an encompassing dynamic config object for this extension. Some suggested names for this that better indicate it's purpose `KubernetesTaskRunnerDynamicConfig` `KubernetesPeonDynamicConfig` `KubernetesTaskExecutionConfig`

arunramani commented:
If we want `Selector` to be general purpose, we need to clean it up a bit. When implementing a selector, you need to have 3 things: the list of selections, the evaluation criteria and the selection key. For this case, we could simplify it to just the list of selections and the selection key. The evaluation criteria can be defaulted to a simple "AND all of the keys except the select key". An example of how it should look
```
"behaviorStrategy": {
    "type": "default",
    "selectKey": "category",
    "categorySelectors": [
      {
        "category": "low-throughput",
        "context.tags": {
          "billingCategory": [
            "streaming_ingestion"
          ]
        },
        "task": {
          "datasource": [
            "wikipedia"
          ]
        }
      },
      {
        "category": "medium-throughput",
        "task": {
          "type": [
            "index_kafka"
          ]
        }
      }
    ]
  }
```
So now the evaluator would look at all of the keys EXCEPT `category` and one a match, it will return the category.

What do you think?

YongGang commented:
> If we want `Selector` to be general purpose, we need to clean it up a bit. When implementing a selector, you need to have 3 things: the list of selections, the evaluation criteria and the selection key. For this case, we could simplify it to just the list of selections and the selection key. The evaluation criteria can be defaulted to a simple "AND all of the keys except the select key". An example of how it should look
> 
> ```
> "behaviorStrategy": {
>     "type": "default",
>     "selectKey": "category",
>     "categorySelectors": [
>       {
>         "category": "low-throughput",
>         "context.tags": {
>           "billingCategory": [
>             "streaming_ingestion"
>           ]
>         },
>         "task": {
>           "datasource": [
>             "wikipedia"
>           ]
>         }
>       },
>       {
>         "category": "medium-throughput",
>         "task": {
>           "type": [
>             "index_kafka"
>           ]
>         }
>       }
>     ]
>   }
> ```
> 
> So now the evaluator would look at all of the keys EXCEPT `category` and one a match, it will return the category.
> 
> What do you think?

I think if fully implement this proposal, it needs to be on reflection based, otherwise still need to preknown what fields to look at. Given the concerns we have on reflection (e.g. perf overhead, not safe etc) and we don't expect some very different fields to look at in the near future so I think the current solution is good enough.

YongGang commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/PodTemplateTaskAdapter.java:
```diff
@@ -126,7 +132,19 @@ public PodTemplateTaskAdapter(
+    }
+    String category = behaviorStrategy.getTaskCategory(task);
+    PodTemplate podTemplate = templates.getOrDefault(
+        category != null ? category : task.getType(),
```
My thinking is when switching to use the new `DynamicTaskExecutionBehaviorStrategy` config, we need to ensure that any existing configs related to task type templates mapping are also considered. Sometimes, these mappings can be overlooked. Therefore, the new strategy should still default to the existing behavior when no category is found, thus facilitating a smoother transition to the new configuration.

georgew5656 replied:
but if you leave this in then users will start to rely on this behavior, i think it would be better to remove it before anyone has tried using the DynamicTaskExecutionBehaviorStrategy

YongGang replied:
updated

georgew5656 commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/KubernetesPeonClient.java:
```diff
@@ -274,6 +279,14 @@ private void emitK8sPodMetrics(Task task, String metric, long durationMs)
     ServiceMetricEvent.Builder metricBuilder = new ServiceMetricEvent.Builder();
     IndexTaskUtils.setTaskDimensions(metricBuilder, task);
+    ExecutionConfig executionConfig = executionConfigRef.get();
+    if (executionConfig != null && executionConfig.getBehaviorStrategy() != null) {
```
would this ever actually be null? with the default can't we set a @Nonnull flag on it and assume it'll be there?

georgew5656 commented on docs/development/extensions-contrib/k8s-jobs.md:
```diff
@@ -217,6 +217,66 @@ data:
+
+We define two categories in the configuration—`low-throughput` and `medium-throughput`—each associated with specific task conditions.
+
+- Low Throughput Category: Tasks that have a context tag `billingCategory=streaming_ingestion` and a datasource of `wikipedia` will be classified under the `low-throughput` category. This classification directs such tasks to utilize a predefined pod template optimized for low throughput requirements.
```
can we make it a bit more clear in the docs that the categorySelectors are ordered, e.g. the first mapping has the highest priority?

suneet-s reviewed (changes requested):
-1 on this change because I do not understand the use of the new interfaces. It seems like what we want here is a config that is scoped to the pod template adapter, but this PR introduces configs for general use across all of the k8s extension. Because of this the names of the interfaces and their uses are not clear to me eg. ExecutionConfig that returns an ExecutionBehaviorStrategy that gets a "category" from the task. It seems like all we need here is something like a `PodTemplateNamingStrategy` or a `PodTemplateSelector`

The other thing that feels clunky is the selector class. It should be an interface so that it can be extended in the future. The current selector class implementation does not provide good errors to users if they reference a field that is not currently supported - like group id. It feels like the Selector class is trying to implement a `Predicate<Task>`

I'd recommend introducing an interface called `PodTemplateSelector` that returns a PodTemplate given a Task object (similar to the BehaviorSelector classes introduced in this patch). For the `Selectors` - I'd recommend renaming them to `Matchers` that implement `Predicate<Task>`. We could then introduce `and`, `not`, `or` matchers and matchers that match on dataSource, tags, any context, task type, etc. The config would then look like

```
"podTemplateSelector" : {
  "type": matcherBasedSelector,
  "templateMatchers": [
    {
      "template": "template0",
      "matcher": {
        "type": or,
        "matchers": [
          {
            "type": "dataSource",
            "matchingNames": ["ds0"]
          },
          {
            "type": "context",
            "field": "myContextKey"
            "matchingNames": ["anyValue"]
          },
        ]
    },
    {
      "template": "template1",
      "matcher": {
        "type": "taskType",
        "matchingTypes": ["index_kafka", "index_kinesis"]
      }
  ]
}
```

YongGang commented:
> -1 on this change because I do not understand the use of the new interfaces. It seems like what we want here is a config that is scoped to the pod template adapter, but this PR introduces configs for general use across all of the k8s extension. Because of this the names of the interfaces and their uses are not clear to me eg. ExecutionConfig that returns an ExecutionBehaviorStrategy that gets a "category" from the task. It seems like all we need here is something like a `PodTemplateNamingStrategy` or a `PodTemplateSelector`
> 
> The other thing that feels clunky is the selector class. It should be an interface so that it can be extended in the future. The current selector class implementation does not provide good errors to users if they reference a field that is not currently supported - like group id. It feels like the Selector class is trying to implement a `Predicate<Task>`
> 
> I'd recommend introducing an interface called `PodTemplateSelector` that returns a PodTemplate given a Task object (similar to the BehaviorSelector classes introduced in this patch). For the `Selectors` - I'd recommend renaming them to `Matchers` that implement `Predicate<Task>`. We could then introduce `and`, `not`, `or` matchers and matchers that match on dataSource, tags, any context, task type, etc. The config would then look like
> 
> ```
> "podTemplateSelectorStrategy" : {
>   "type": matcherBased,
>   "templateMatchers": [
>     {
>       "template": "template0",
>       "matcher": {
>         "type": or,
>         "matchers": [
>           {
>             "type": "dataSource",
>             "matchingNames": ["ds0"]
>           },
>           {
>             "type": "context",
>             "field": "myContextKey"
>             "matchingNames": ["anyValue"]
>           },
>         ]
>       }
>     },
>     {
>       "template": "template1",
>       "matcher": {
>         "type": "taskType",
>         "matchingTypes": ["index_kafka", "index_kinesis"]
>       }
>     }
>   ]
> }
> ```

@suneet-s I have addressed most of the comments except for the `Matchers` proposal. While it’s a good suggestion, I don’t believe it’s the right fit for our scenario for a couple of reasons:
1. **Complexity and Applicability of Operators**: `Matchers` provide a variety of operators such as `and`, `not`, and `or` for evaluation. Our current design with sequence-based selectors is already complex and requires careful attention to detail. Introducing additional operators would further complicate the understanding of the dynamic config. Specifically, the `not` and `or` operators might not be very useful for our focused criteria of template selection. For example, the `not` operator could yield unexpected results, especially as new Task types are continuously added, potentially covering unintended scenarios without careful usage.
2. **Relevance of Typed Matchers**: While the typed `Matcher` approach is excellent for extensibility and appears future-proof, it is more aligned with scenarios like Druid’s Filter, where many implementations exist for type safety and performance enhancements. In our case, however, the field comparisons are predominantly string-based. Given that our main input in this K8s extension is a **Task**, we do not anticipate needing significantly different criteria from what is currently covered. Should future requirements drastically diverge, adopting a new strategy would be more appropriate than extending the current one.

Additionally, implementing `Matchers` would require introducing many small Jackson classes, which could bloat the codebase. Our current implementation of Selector is cleaner and more streamlined in comparison.

suneet-s commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/PodTemplateSelectStrategy.java:
```diff
@@ -0,0 +1,63 @@
+    if (podTemplate != null) {
+      return Pair.of(templateKey, podTemplate);
+    } else {
+      return Pair.of("base", templates.get("base"));
```
This used to throw an ISE if `base` did not exist

YongGang replied:
We still have this logic in `PodTemplateTaskAdapter` which also make sure there is base template configured.

suneet-s commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/PodTemplateSelectStrategy.java:
```diff
@@ -0,0 +1,63 @@
+   * allows for customized resource allocation and management tailored to the task's specific requirements.
+   *
+   * @param task The task for which the Pod template is determined.
+   * @return A key-value pair representing the selected Pod template. If no matching template is found,
+   *         the method falls back to a base template.
```
What if no base template is found?

I think the interface should probably throw an exception if it can not find a template for the task.

suneet-s commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/PodTemplateSelectStrategy.java:
```diff
@@ -0,0 +1,63 @@
+   */
+  Pair<String, PodTemplate> getPodTemplateForTask(Task task, Map<String, PodTemplate> templates);
+
+  default Pair<String, PodTemplate> getTemplateOrDefault(String templateKey, Map<String, PodTemplate> templates)
```
I do not think this should be part of the interface definition. To a caller, it is not clear whether to call `getPodTemplateForTask` or `getTemplateOrDefault`

suneet-s commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/KubernetesTaskExecutionConfigResource.java:
```diff
@@ -0,0 +1,159 @@
+ * <p>This class handles the CRUD operations for execution configurations and provides
+ * endpoints to update, retrieve, and manage the history of these configurations.</p>
+ */
+@Path("/druid/indexer/v1/k8s/taskRunner")
+public class KubernetesTaskExecutionConfigResource
```
```suggestion
@Path("/druid/indexer/v1/k8s/taskRunner/executionConfig")
public class KubernetesTaskExecutionConfigResource
```

So that the class name matches the URL, then the functions should only specify the `@Path` annotation if they need a path under `.../executionConfig`

suneet-s reviewed (changes requested):
Thanks for the updates @YongGang. I have reviewed the src/main classes. 

I am still -1 on this change because of the way the selector class is written. 
The Selector contains a `selection key` and implicit matching conditions. These matching conditions are specific to a usecase, but it is not clear why these specific fields are chosen. IMO this poses a UX problem for people adopting this feature. The other issue with this approach is that we are not able to provide good user validation when a user passes in a key to the task map that is not currently supported.

Please separate the selection key from what is being used to match the task with a selection key into a separate object aka instead of
``` 
{
  "selectionKey": "someKey",
  "context.tags": {..},
  "task": {...}
}
```

make it matching part it's own object
``` 
{
  "selectionKey": "someKey",
  "matcher": {
    "type": "myCustomMatcher",
    "context.tags": {..},
    "task": {...}
  }
}
```
In the example above, I can not think of a good name that would explain why someone would want to use that matcher, which is why I am proposing smaller matchers - like a `DataSourceMatcher`, `TaskTypeMatcher`, `TagsMatcher`, etc.

Regarding your concerns about smaller, more composable matchers:

1. RE: Complexity of operators: We do not need to introduce any matchers that make it harder to understand what the rules are trying to say. The minimum work needed in this patch is to provide an `and` matcher, a `tags` matcher, a `dataSource` matcher and a `taskType` matcher. It is no more complex than the current proposal with these 4 matchers.
2. RE: Relevance of Typed matchers: I do not understand this concern.
3. RE: small jackson classes / code bloat: Smaller classes are easier to test and maintain as there is less logic in it. I do not understand the code bloat concern.


Also, please limit the use of nulls in the code. Please try to make everything non-null as this makes it less likely to introduce null handling bugs. If something really needs to be nullable, consider using Optionals as it forces the devs and the reviewers to think about what to do when the optional is absent.

YongGang commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/DynamicTaskPodTemplateSelectStrategy.java:
```diff
@@ -0,0 +1,103 @@
+      templateKey = templateSelectors.stream()
+                                     .filter(selector -> selector.evaluate(task))
+                                     .findFirst()
+                                     .map(Selector::getSelectionKey)
```
Since the Selector class will be reused for other purpose, that's why it's called SelectionKey instead of a specific TemplateKey see my reply here https://github.com/apache/druid/pull/16510#discussion_r1623791655

YongGang commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/execution/DynamicTaskPodTemplateSelectStrategy.java:
```diff
@@ -0,0 +1,103 @@
+                                     .filter(selector -> selector.evaluate(task))
+                                     .findFirst()
+                                     .map(Selector::getSelectionKey)
+                                     .orElse(null);
```
The fallback to base template is implemented in `PodTemplateSelectStrategy` interface, so no need to have this logic in each strategy implementation.

YongGang commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/KubernetesPeonClient.java:
```diff
@@ -270,10 +270,16 @@ Pod getPeonPodWithRetries(KubernetesClient client, String jobName, int quietTrie
+          "podTemplate",
+          job.getMetadata().getAnnotations().get(DruidK8sConstants.POD_TEMPLATE_KEY)
+      );
+    }
```
It's for metrics reporting, but don't think we need that anymore, will revert this part.

YongGang commented:
Thanks @suneet-s for your comments:

> The Selector contains a selection key and implicit matching conditions. These matching conditions are specific to a usecase, but it is not clear why these specific fields are chosen.

If we rename `Selector` to `TaskPropertiesSelector` should address this concern?

For this following suggestion:
```
{
  "selectionKey": "someKey",
  "matcher": {
    "type": "myCustomMatcher",
    "context.tags": {..},
    "task": {...}
  }
}
```
are we actually suggesting to do this?:
```
{
  "selectionKey": "someKey",
  "matchers": [
    {
      "type": "dataSource",
      "matchingNames": ["ds0"]
    },
    {
      "type": "context",
      "field": "myContextKey"
      "matchingNames": ["anyValue"]
    },
  ]
}
```
Otherwise if we still have a single `matcher` within a `selector`, I'm not sure about the benefit of introducing another layer of Jackson object.

YongGang commented on extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesOverlordModule.java:
```diff
@@ -75,6 +79,7 @@ public void configure(Binder binder)
     JsonConfigProvider.bind(binder, IndexingServiceModuleHelper.INDEXER_RUNNER_PROPERTY_PREFIX, KubernetesTaskRunnerConfig.class);
     JsonConfigProvider.bind(binder, K8SANDWORKER_PROPERTIES_PREFIX, KubernetesAndWorkerTaskRunnerConfig.class);
     JsonConfigProvider.bind(binder, "druid.indexer.queue", TaskQueueConfig.class);
+    JacksonConfigProvider.bind(binder, KubernetesTaskRunnerDynamicConfig.CONFIG_KEY, KubernetesTaskRunnerDynamicConfig.class, null);
```
As the binding is to a record from database, the record can be updated or deleted at runtime, thus we still need to do some null checkings to be safe. It's also similar to how `WorkerBehaviorConfig` is used now and the the null checking is only in one place, so it's quite clean.

YongGang commented:
@suneet-s I have addressed your comments, please take a look. Thanks.

suneet-s reviewed (approved):
Thanks for incorporating the suggestions @YongGang
